File: eeg-backend/neuro_yogic/feature_extractor.py
"""
feature_extractor.py
====================
Enhanced signal processing pipeline based on:
"Electroencephalographic Mapping of Yogic Physiology: An Integrated
Neuroscientific Framework for Real-Time Biofeedback Calibration"

Band definitions aligned with paper specifications:
  delta     : 0.5–4 Hz   — Tamas / deep sleep / Mudha bhumi
  theta     : 4–8 Hz     — Sattva / creativity / Ekagra bhumi (Fm-θ)
  alpha     : 8–13 Hz    — Sattva / relaxed awareness / Vikshipta bhumi
  low_beta  : 13–18 Hz   — SMR/mid-beta — neutral to calm-focus
  high_beta : 18–30 Hz   ← PRIMARY Rajas / Kshipta marker (paper: "desynchronized
                           low-amplitude High Beta 18-30 Hz in prefrontal areas")
  gamma     : 30–50 Hz   — Niruddha / peak states / Gunatita

New derived metrics:
  faa : ln(alpha_right) − ln(alpha_left)
        Standard Frontal Alpha Asymmetry (neuroscience convention).
        Positive → left PFC activation → Pingala (FAA > +0.15 per paper)
        Negative → right PFC activation → Ida   (FAA < −0.15 per paper)
        Near zero → Sushumna equilibrium         (|FAA| ≤ 0.15 per paper)

  plv : Phase Locking Value between left & right alpha channels.
        Measures interhemispheric synchrony — the neural signature of Sushumna
        and Niruddha states. PLV > 0.80 = high coherence (paper: Niruddha).

Electrode layout (4-channel Muse S / Muse 2):
  Left  channels: indices 0, 1  (TP9, AF7)
  Right channels: indices 2, 3  (AF8, TP10)
"""

from typing import List, Optional, Tuple

import numpy as np
from scipy.signal import butter, iirnotch, sosfiltfilt, tf2sos, welch, hilbert

from neuro_yogic.aperiodic import aperiodic_fit
from neuro_yogic.artifact import artifact_screen
from neuro_yogic.complexity import complexity_features
from neuro_yogic.connectivity import connectivity_features
from neuro_yogic.iaf import individual_alpha_freq

# ── Band definitions (paper-aligned) ─────────────────────────────────────────
EEG_BANDS = {
    "delta":     (0.5,  4.0),
    "theta":     (4.0,  8.0),
    "alpha":     (8.0, 13.0),
    "low_beta":  (13.0, 18.0),   # SMR + mid-beta (calm-focus, Manipura/Vishuddha)
    "high_beta": (18.0, 30.0),   # Rajas marker (Kshipta / stress / fight-flight)
    "gamma":     (30.0, 50.0),   # Niruddha / peak states
}

# Feature vector column names (10-D)
FEATURE_COLUMNS = [
    "delta", "theta", "alpha", "low_beta", "high_beta", "gamma",
    "alpha_left", "alpha_right", "faa", "plv",
]

# NumPy 2.x renamed np.trapz -> np.trapezoid; support both.
_trapz = getattr(np, "trapezoid", None) or getattr(np, "trapz")


class FeatureExtractor:
    """
    Transforms raw multi-channel EEG into the 10-D classifier feature vector.

    New vs. v1:
    • Splits beta into low_beta (13-18 Hz) and high_beta (18-30 Hz) — critical
      for accurate Rajas detection (the paper specifies HIGH beta as Rajas).
    • Computes FAA (ln-ratio Frontal Alpha Asymmetry) — the standard metric for
      Nadi/Swara lateralization used in clinical neuroscience.
    • Computes PLV (Phase Locking Value) between left/right alpha channels —
      the neural signature of interhemispheric coherence (Sushumna / Niruddha).
    """

    def __init__(
        self,
        sample_rate:   int                = 256,
        left_indices:  Optional[List[int]] = None,
        right_indices: Optional[List[int]] = None,
        notch_freq:    float              = 50.0,
    ) -> None:
        self._sr        = sample_rate
        self._left_idx  = left_indices  or [0, 1]
        self._right_idx = right_indices or [2, 3]
        self._nperseg   = max(sample_rate // 2, 32)   # 0.5-second Welch segments

        nyq = sample_rate / 2.0

        # Butterworth bandpass (0.5–50 Hz), order 4, SOS format
        self._bp_sos = butter(
            4, [0.5 / nyq, min(49.9, 50.0) / nyq],
            btype="bandpass", output="sos"
        )

        # IIR notch (50 or 60 Hz power-line interference)
        b, a = iirnotch(min(notch_freq / nyq, 0.999), Q=30.0)
        self._notch_sos = tf2sos(b, a)

        # Alpha bandpass for PLV computation (8–13 Hz)
        self._alpha_sos = butter(
            4, [8.0 / nyq, 13.0 / nyq],
            btype="bandpass", output="sos"
        )

    # ── Public API ─────────────────────────────────────────────────────────────

    def extract(
        self,
        raw_eeg: np.ndarray,
        meta: Optional[dict] = None,
    ) -> Tuple[np.ndarray, dict]:
        """
        Run the full pipeline on one EEG epoch.

        Parameters
        ----------
        raw_eeg : ndarray (n_channels, n_samples)  — raw µV data
        meta    : optional dict (e.g. {"is_padded": True})

        Returns
        -------
        features : ndarray (10,) — [delta, theta, alpha, low_beta, high_beta,
                                    gamma, alpha_left_rel, alpha_right_rel, faa, plv]
        info     : dict — all intermediate products for downstream classifiers
        """
        is_padded = (meta or {}).get("is_padded", False)
        raw_eeg   = np.atleast_2d(np.array(raw_eeg, dtype=np.float64))
        n_ch, n_samp = raw_eeg.shape

        # ── Step 0: Artifact / blink screening (on the raw signal) ────────────
        artifact = artifact_screen(raw_eeg)

        # ── Step 1: Bandpass filter (0.5–50 Hz) ───────────────────────────────
        filtered = sosfiltfilt(self._bp_sos, raw_eeg, axis=1)

        # ── Step 2: Notch filter (power-line) ─────────────────────────────────
        filtered = sosfiltfilt(self._notch_sos, filtered, axis=1)

        # ── Step 3: Welch PSD ─────────────────────────────────────────────────
        # freqs: (n_freq,)   psd: (n_channels, n_freq)
        freqs, psd = welch(
            filtered,
            fs       = self._sr,
            nperseg  = min(self._nperseg, n_samp),
            noverlap = min(self._nperseg, n_samp) // 2,
            axis     = 1,
        )

        # ── Step 4: Band power integration (trapezoid rule) ───────────────────
        band_abs = {}
        for band_name, (lo, hi) in EEG_BANDS.items():
            mask         = (freqs >= lo) & (freqs < hi)
            if not mask.any():
                band_abs[band_name] = 0.0
                continue
            power_per_ch = _trapz(psd[:, mask], freqs[mask], axis=1)
            band_abs[band_name] = float(np.mean(power_per_ch))

        # ── Step 5: Relative power normalisation ──────────────────────────────
        total    = sum(band_abs.values()) or 1e-10
        band_rel = {k: v / total for k, v in band_abs.items()}

        # Also expose a legacy "beta" key = low_beta + high_beta combined
        band_rel["beta"] = band_rel["low_beta"] + band_rel["high_beta"]

        # ── Step 6: Hemispheric Alpha Asymmetry ───────────────────────────────
        alpha_mask = (freqs >= 8.0) & (freqs < 13.0)

        def _mean_alpha(ch_idx: List[int]) -> float:
            valid = [i for i in ch_idx if i < n_ch]
            if not valid:
                return 1e-12
            vals = _trapz(psd[valid][:, alpha_mask], freqs[alpha_mask], axis=1)
            return float(max(np.mean(vals), 1e-12))

        alpha_left  = _mean_alpha(self._left_idx)
        alpha_right = _mean_alpha(self._right_idx)
        asymmetry   = alpha_right - alpha_left   # raw difference (legacy)

        # Standard FAA: ln(alpha_right) − ln(alpha_left)
        # This is the conventional neuroscience formula — it normalises for
        # overall alpha level and produces symmetric values around 0.
        try:
            faa = float(np.log(alpha_right) - np.log(alpha_left))
        except (ValueError, ZeroDivisionError):
            faa = 0.0
        faa = float(np.clip(faa, -2.0, 2.0))   # reasonable range for display

        # Relative alpha (for feature vector)
        alpha_left_rel  = alpha_left  / (total + 1e-12)
        alpha_right_rel = alpha_right / (total + 1e-12)

        # ── Step 7: Phase Locking Value (alpha interhemispheric coherence) ────
        # Measures how consistently left and right hemispheres oscillate in phase.
        # High PLV (>0.80) = Sushumna / Niruddha coherence signature.
        plv = self._compute_plv(filtered)

        # ── Step 7b: Non-linear complexity features (deep-state markers) ──────
        # Computed on the band-pass+notch signal, per channel then averaged.
        complexity = complexity_features(filtered)

        # ── Step 7c: Aperiodic (1/f) decomposition ───────────────────────────
        # Parameterise the non-oscillatory spectral background over 2–40 Hz.
        aperiodic = aperiodic_fit(freqs, psd)

        # ── Step 7d: Non-linear connectivity (wSMI) ──────────────────────────
        # Weighted symbolic mutual information over channel pairs.
        connectivity = connectivity_features(filtered)

        # ── Step 7e: Individual Alpha Frequency (spectral centroid 7–13 Hz) ──
        iaf = individual_alpha_freq(freqs, psd)

        # ── Step 8: Assemble 10-D feature vector ──────────────────────────────
        features = np.array([
            band_rel["delta"],
            band_rel["theta"],
            band_rel["alpha"],
            band_rel["low_beta"],
            band_rel["high_beta"],
            band_rel["gamma"],
            alpha_left_rel,
            alpha_right_rel,
            faa,
            plv,
        ], dtype=np.float64)

        info = {
            "band_absolute":   band_abs,
            "band_relative":   band_rel,
            "alpha_left":      alpha_left,
            "alpha_right":     alpha_right,
            "alpha_asymmetry": asymmetry,
            "faa":             faa,
            "plv":             plv,
            "complexity":      complexity,
            "aperiodic":       aperiodic,
            "connectivity":    connectivity,
            "iaf":             iaf,
            "artifact_flagged": artifact["artifact_flagged"],
            "signal_quality":   artifact["signal_quality"],
            "gamma_spike":     band_rel["gamma"] > 0.12,
            "is_padded":       is_padded,
        }
        return features, info

    # ── Private helpers ────────────────────────────────────────────────────────

    def _compute_plv(self, filtered: np.ndarray) -> float:
        """
        Compute alpha-band Phase Locking Value between left and right channels.

        PLV = |mean(exp(i * Δφ(t)))| where Δφ = φ_right − φ_left.
        Range: 0 (no coherence) → 1.0 (perfect phase lock).

        Returns 0.5 (neutral) if insufficient channels or samples.
        """
        n_ch, n_samp = filtered.shape
        left_idx  = [i for i in self._left_idx  if i < n_ch]
        right_idx = [i for i in self._right_idx if i < n_ch]
        if not left_idx or not right_idx or n_samp < 64:
            return 0.5

        try:
            # Alpha-bandpass the signals
            left_alpha  = sosfiltfilt(self._alpha_sos, filtered[left_idx[0]],  axis=0)
            right_alpha = sosfiltfilt(self._alpha_sos, filtered[right_idx[0]], axis=0)

            # Analytic signal → instantaneous phase
            left_phase  = np.angle(hilbert(left_alpha))
            right_phase = np.angle(hilbert(right_alpha))

            phase_diff = right_phase - left_phase
            plv = float(np.abs(np.mean(np.exp(1j * phase_diff))))
            return float(np.clip(plv, 0.0, 1.0))
        except Exception:
            return 0.5

File: eeg-backend/src/NeuroYogic.SignalProcessing/FeatureExtractor.cs
using NeuroYogic.Domain.Analysis;
using NeuroYogic.SignalProcessing.Dsp;

namespace NeuroYogic.SignalProcessing;

/// <summary>
/// Faithful C# port of the Python <c>FeatureExtractor</c>: band-pass + notch
/// filtering, Welch PSD, band-power integration, hemispheric alpha asymmetry
/// (ln-ratio FAA), and alpha-band interhemispheric PLV.
/// </summary>
public sealed class FeatureExtractor : IFeatureExtractor
{
    private readonly int _defaultLeftIdx0;
    private readonly int _defaultLeftIdx1;
    private readonly int _defaultRightIdx0;
    private readonly int _defaultRightIdx1;
    private readonly double _notchFreq;

    public FeatureExtractor(
        IReadOnlyList<int>? leftIndices = null,
        IReadOnlyList<int>? rightIndices = null,
        double notchFreq = 50.0)
    {
        leftIndices ??= new[] { 0, 1 };
        rightIndices ??= new[] { 2, 3 };
        _defaultLeftIdx0 = leftIndices[0];
        _defaultLeftIdx1 = leftIndices.Count > 1 ? leftIndices[1] : leftIndices[0];
        _defaultRightIdx0 = rightIndices[0];
        _defaultRightIdx1 = rightIndices.Count > 1 ? rightIndices[1] : rightIndices[0];
        _notchFreq = notchFreq;
    }

    public FeatureSet Extract(double[][] rawEeg, int sampleRate = 256, bool isPadded = false)
    {
        if (rawEeg.Length == 0) throw new ArgumentException("EEG data has no channels.", nameof(rawEeg));
        var nCh = rawEeg.Length;
        var nSamp = rawEeg[0].Length;

        var nyq = sampleRate / 2.0;
        var nperseg = Math.Max(sampleRate / 2, 32);

        // ── Step 0: artifact / blink screening (on the raw signal) ──
        var (artifactFlagged, signalQuality) = ArtifactRejection.Screen(rawEeg);

        var leftIdx = new[] { _defaultLeftIdx0, _defaultLeftIdx1 };
        var rightIdx = new[] { _defaultRightIdx0, _defaultRightIdx1 };

        // ── Filter designs (match Python constructor) ──
        var bpSos = Butterworth.BandpassSos(4, 0.5 / nyq, Math.Min(49.9, 50.0) / nyq);
        var notchSos = IirNotch.Sos(Math.Min(_notchFreq / nyq, 0.999), 30.0);
        var alphaSos = Butterworth.BandpassSos(4, EegBands.AlphaLow / nyq, EegBands.AlphaHigh / nyq);

        // ── Step 1+2: band-pass then notch, per channel ──
        var filtered = new double[nCh][];
        for (var c = 0; c < nCh; c++)
        {
            var bp = SosFilter.FiltFilt(bpSos, rawEeg[c]);
            filtered[c] = SosFilter.FiltFilt(notchSos, bp);
        }

        // ── Step 3: Welch PSD per channel ──
        var welchN = Math.Min(nperseg, nSamp);
        var noverlap = welchN / 2;
        double[] freqs = Array.Empty<double>();
        var psd = new double[nCh][];
        for (var c = 0; c < nCh; c++)
        {
            var res = Welch.Compute(filtered[c], sampleRate, welchN, noverlap);
            freqs = res.Frequencies;
            psd[c] = res.Psd;
        }

        // ── Step 4: band-power integration (trapezoid), mean over channels ──
        double BandAbs(double lo, double hi)
        {
            var mask = MaskIndices(freqs, lo, hi);
            if (mask.Count == 0) return 0.0;
            double sum = 0.0;
            for (var c = 0; c < nCh; c++) sum += TrapzMasked(psd[c], freqs, mask);
            return sum / nCh;
        }

        var deltaAbs = BandAbs(0.5, 4.0);
        var thetaAbs = BandAbs(4.0, 8.0);
        var alphaAbs = BandAbs(8.0, 13.0);
        var lowBetaAbs = BandAbs(13.0, 18.0);
        var highBetaAbs = BandAbs(18.0, 30.0);
        var gammaAbs = BandAbs(30.0, 50.0);

        var absolute = new BandPowers(deltaAbs, thetaAbs, alphaAbs, lowBetaAbs, highBetaAbs, gammaAbs);

        // ── Step 5: relative power ──
        var total = deltaAbs + thetaAbs + alphaAbs + lowBetaAbs + highBetaAbs + gammaAbs;
        if (total <= 0) total = 1e-10;
        var relative = new BandPowers(
            deltaAbs / total, thetaAbs / total, alphaAbs / total,
            lowBetaAbs / total, highBetaAbs / total, gammaAbs / total);

        // ── Step 6: hemispheric alpha asymmetry ──
        var alphaMask = MaskIndices(freqs, 8.0, 13.0);

        double MeanAlpha(int[] idx)
        {
            var valid = idx.Where(i => i < nCh).ToArray();
            if (valid.Length == 0) return 1e-12;
            double sum = 0.0;
            foreach (var i in valid) sum += TrapzMasked(psd[i], freqs, alphaMask);
            return Math.Max(sum / valid.Length, 1e-12);
        }

        var alphaLeft = MeanAlpha(leftIdx);
        var alphaRight = MeanAlpha(rightIdx);
        var asymmetry = alphaRight - alphaLeft;

        double faa;
        try
        {
            faa = Math.Log(alphaRight) - Math.Log(alphaLeft);
        }
        catch
        {
            faa = 0.0;
        }
        faa = Math.Clamp(faa, -2.0, 2.0);

        // ── Step 7: PLV ──
        var plv = ComputePlv(filtered, alphaSos, leftIdx, rightIdx);

        // ── Step 7b: non-linear complexity features (deep-state markers) ──
        var complexity = Complexity.Compute(filtered);

        // ── Step 7c: aperiodic (1/f) decomposition ──
        var aperiodic = Aperiodic.Fit(freqs, psd);

        // ── Step 7d: non-linear connectivity (wSMI) ──
        var connectivity = Connectivity.Compute(filtered);

        // ── Step 7e: Individual Alpha Frequency ──
        var iaf = Iaf.Estimate(freqs, psd);

        return new FeatureSet
        {
            BandRelative = relative,
            BandAbsolute = absolute,
            AlphaLeft = alphaLeft,
            AlphaRight = alphaRight,
            AlphaAsymmetry = asymmetry,
            Faa = faa,
            Plv = plv,
            Complexity = complexity,
            Aperiodic = aperiodic,
            Connectivity = connectivity,
            Iaf = iaf,
            ArtifactFlagged = artifactFlagged,
            SignalQuality = signalQuality,
            GammaSpike = relative.Gamma > 0.12,
            IsPadded = isPadded,
        };
    }

    private static double ComputePlv(double[][] filtered, double[][] alphaSos, int[] leftIdx, int[] rightIdx)
    {
        var nCh = filtered.Length;
        var nSamp = filtered[0].Length;
        var left = leftIdx.Where(i => i < nCh).ToArray();
        var right = rightIdx.Where(i => i < nCh).ToArray();
        if (left.Length == 0 || right.Length == 0 || nSamp < 64) return 0.5;

        try
        {
            var leftAlpha = SosFilter.FiltFilt(alphaSos, filtered[left[0]]);
            var rightAlpha = SosFilter.FiltFilt(alphaSos, filtered[right[0]]);
            var leftPhase = Hilbert.InstantaneousPhase(leftAlpha);
            var rightPhase = Hilbert.InstantaneousPhase(rightAlpha);

            double sumRe = 0.0, sumIm = 0.0;
            for (var i = 0; i < nSamp; i++)
            {
                var d = rightPhase[i] - leftPhase[i];
                sumRe += Math.Cos(d);
                sumIm += Math.Sin(d);
            }
            var plv = Math.Sqrt(sumRe * sumRe + sumIm * sumIm) / nSamp;
            return Math.Clamp(plv, 0.0, 1.0);
        }
        catch
        {
            return 0.5;
        }
    }

    /// <summary>Indices where lo ≤ freq &lt; hi (matches Python mask semantics).</summary>
    private static List<int> MaskIndices(double[] freqs, double lo, double hi)
    {
        var idx = new List<int>();
        for (var i = 0; i < freqs.Length; i++)
            if (freqs[i] >= lo && freqs[i] < hi) idx.Add(i);
        return idx;
    }

    /// <summary>Trapezoidal integral of y over x at the given (contiguous) mask indices.</summary>
    private static double TrapzMasked(double[] y, double[] x, List<int> mask)
    {
        if (mask.Count < 2) return 0.0;
        double sum = 0.0;
        for (var k = 0; k < mask.Count - 1; k++)
        {
            var i0 = mask[k];
            var i1 = mask[k + 1];
            sum += (x[i1] - x[i0]) * (y[i1] + y[i0]) / 2.0;
        }
        return sum;
    }
}
